tests/config_test.py

```diff
@@ -1040,8 +1036,84 @@ def test_migrate_from_13_to_14_with_gradle_and_jdk8_ask_user_enter_valid(self, m
+        sample_config = {
+            "meta": {
+                "config.version": 14
+            }
+        }
+        config_file.store(sample_config)
+        config.migrate(config_file, 14, 15, out=null_output)
+
+        self.assertTrue(config_file.backup_created)
+        self.assertEqual("15", config_file.config["meta"]["config.version"])
+
+    def test_migrate_from_14_to_15_with_gradle(self):
+        config_file = InMemoryConfigStore("test")
+        sample_config = {
+            "meta": {
+                "config.version": 14
+            },
+            "build": {
+                "gradle.bin": "/usr/local/bin/gradle"
+            },
+            "runtime": {
+                "java.home": "/usr/lib/java8",
+                "java10.home": "/usr/lib/java10"
+            }
+        }
+        config_file.store(sample_config)
+        config.migrate(config_file, 14, 15, out=null_output)
+
+        self.assertTrue(config_file.backup_created)
+        self.assertEqual("15", config_file.config["meta"]["config.version"])
+        self.assertNotIn("build", config_file.config)
+
+    @mock.patch("esrally.utils.io.guess_java_home")
+    @mock.patch("esrally.utils.jvm.is_early_access_release")
+    @mock.patch("esrally.utils.jvm.major_version")
+    def test_migrate_from_14_to_15_with_source_plugin_definition(self, major_version, is_early_access_release, guess_java_home):
+        guess_java_home.return_value = None
+        is_early_access_release.return_value = False
+        major_version.return_value = 10
+
+        config_file = InMemoryConfigStore("test")
+        sample_config = {
+            "meta": {
+                "config.version": 14
+            },
+            "build": {
+                "gradle.bin": "/usr/local/bin/gradle"
+            },
+            "runtime": {
+                "java.home": "/usr/lib/java10",
+                "java10.home": "/usr/lib/java10"
+            },
+            "source": {
+                "plugin.x-pack.remote.repo.url": "git@github.com:elastic/x-pack-elasticsearch.git",
+                "plugin.x-pack.src.subdir": "elasticsearch-extra/x-pack-elasticsearch",
+                "plugin.x-pack.build.task": ":x-pack-elasticsearch:plugin:assemble",
+                "plugin.x-pack.build.artifact.subdir": "plugin/build/distributions"
+            }
+        }
+        config_file.store(sample_config)
+        string_buffer = StringIO()
+        config.migrate(config_file, 14, 15, out=string_buffer.write)
+        string_buffer.seek(0)
+
+        self.assertTrue(config_file.backup_created)
+        self.assertEqual("15", config_file.config["meta"]["config.version"])
+        self.assertNotIn("build", config_file.config)
+        self.assertIn(
+            "You will need to edit the plugin [{}] section in your rally ini file and change from:\n"
+            "                          "
+            "[{} = {}] to [{} = <the full command>].".format("x-pack",
```

Comment:
How about we provide the fully rendered string here in order to improve readability and also to ensure we did not mess up formatting unintentionally?

Reply:
Yes makes sense. Addressing.